runtime/hash.c

```diff
@@ -237,12 +230,27 @@ CAMLprim value caml_hash(value count, value limit, value seed, value obj)
         h = caml_hash_mix_uint32(h, Infix_offset_val(v));
         v = v - Infix_offset_val(v);
         goto again;
+      case Closure_tag:
+        /* !! something complicated */
+        #if 0
+        /* v is a pointer outside the heap, probably a code pointer.
+           Shall we count it?  Let's say yes by compatibility with old code. */
+        h = caml_hash_mix_intnat(h, v);
+        num--;
+        #endif
+        h = 42;
+        break;
+      case Cont_tag:
+        /* All continuations hash to the same value,
+           since we have no idea how to distinguish them. */
+        h = 42;
```

Comment:
Also, by setting `h = 42` you forget the result of all the hashing previously done!  Just don't update `h`.